src/eckit/log/Log.cc

```diff
@@ -99,105 +105,86 @@ std::ostream& Log::message() {
-struct CreateLogChannel {
-
-    virtual Channel* createChannel() = 0;
-
-    Channel* operator()() {
-        try {
-            return createChannel();
-        }
-        catch (std::exception& e) {
-            std::cerr << "Exception caught when creating channel: " << e.what() << std::endl;
-            return new Channel(new OStreamTarget(std::cout));
-        }
-    }
-};
-
-struct CreateMetricsChannel : public CreateLogChannel {
-    virtual Channel* createChannel() { return new Channel(Main::instance().createMetricsLogTarget()); }
-};
-
 Channel& Log::metrics() {
     if (!Main::ready()) {
-        static Channel empty(new PrefixTarget("PRE-MAIN-METRICS", new OStreamTarget(std::cout)));
-        return empty;
+        thread_local Channel preMainMetrics(new PrefixTarget("PRE-MAIN-METRICS", new OStreamTarget(std::cout)));
+        return preMainMetrics;
     }
-    static ThreadSingleton<Channel, CreateMetricsChannel> x;
-    return x.instance();
-}
 
+    if (Main::finalised()) {
+        thread_local Channel postMainMetrics(new PrefixTarget("POST-MAIN-METRICS", new OStreamTarget(std::cout)));
+        return postMainMetrics;
+    }
 
-struct CreateInfoChannel : public CreateLogChannel {
-    virtual Channel* createChannel() { return new Channel(Main::instance().createInfoLogTarget()); }
-};
+    thread_local Channel mainMetrics(Main::instance().createMetricsLogTarget());
+    return mainMetrics;
+}
 
 Channel& Log::info() {
     if (!Main::ready()) {
-        static Channel empty(new PrefixTarget("PRE-MAIN-INFO", new OStreamTarget(std::cout)));
-        return empty;
+        thread_local Channel preMainInfo(new PrefixTarget("PRE-MAIN-INFO", new OStreamTarget(std::cout)));
+        return preMainInfo;
     }
-    static ThreadSingleton<Channel, CreateInfoChannel> x;
-    return x.instance();
-}
 
-struct CreateErrorChannel : public CreateLogChannel {
-    virtual Channel* createChannel() { return new Channel(Main::instance().createErrorLogTarget()); }
-};
+    if (Main::finalised()) {
+        thread_local Channel postMainInfo(new PrefixTarget("POST-MAIN-INFO", new OStreamTarget(std::cout)));
+        return postMainInfo;
+    }
+
+    thread_local Channel mainInfo(Main::instance().createInfoLogTarget());
+    return mainInfo;
+}
 
 Channel& Log::error() {
     if (!Main::ready()) {
-        static Channel empty(new PrefixTarget("PRE-MAIN-ERROR", new OStreamTarget(std::cout)));
-        return empty;
+        thread_local Channel preMainError(new PrefixTarget("PRE-MAIN-ERROR", new OStreamTarget(std::cout)));
+        return preMainError;
     }
-    static ThreadSingleton<Channel, CreateErrorChannel> x;
-    return x.instance();
-}
 
-struct CreateWarningChannel : public CreateLogChannel {
-    virtual Channel* createChannel() { return new Channel(Main::instance().createWarningLogTarget()); }
-};
+    if (Main::finalised()) {
+        thread_local Channel postMainError(new PrefixTarget("POST-MAIN-ERROR", new OStreamTarget(std::cout)));
```

Comment:
Put in a comment:


This may still cause a segfault due to detruction order in library shutdown - 
We may have to return an emptychannel that won't cause any harm when it has been destructed officially

Reply:
I am hesitant about this comment (not clear to me) because it can be very misleading if we haven't seen a case before

which type of library, runtime or system ?

for example, calling Log::info/debug/error in a system::~Library will fail but nothing to do with what Main::finalised does.